kratos/utilities/assign_mpcs_to_neighbours_utility.cpp

```diff
@@ -0,0 +1,376 @@
+
+    ModelPart::MasterSlaveConstraintType const& r_clone_constraint = KratosComponents<MasterSlaveConstraint>::Get("LinearMasterSlaveConstraint");
+
+    // Create a concurrent queue for constraints
+    using MasterSlaveConstarintQueue = moodycamel::ConcurrentQueue<ModelPart::MasterSlaveConstraintType::Pointer>;
+
+    MasterSlaveConstarintQueue concurrent_constraints;
+
+    // Declare a counter variable outside the loop as std::atomic<int>
+    std::atomic<int> i(0);
+
+    block_for_each(nodes_array, [&](Node<3>::Pointer& pNode) {
+
+        double local_radius = Radius;
+
+        ResultNodesContainerType r_result(0);
+
+        while (r_result.size() < MinNumOfNeighNodes) {
+            r_result.clear();
+            SearchNodesInRadiusForNode(pNode, local_radius, r_result);
+            local_radius += Radius;
+        }
+
+        // Get Dofs and Coordinates
+        DofPointerVectorType r_cloud_of_dofs, r_slave_dof;
+        Matrix r_cloud_of_nodes_coordinates;
+        array_1d<double, 3> r_slave_coordinates;
+        GetDofsAndCoordinatesForNode(pNode, rVariable, r_slave_dof, r_slave_coordinates);
+        GetDofsAndCoordinatesForNodes(r_result, rVariable, r_cloud_of_dofs, r_cloud_of_nodes_coordinates);
+
+        // Calculate shape functions
+        Vector r_n_container;
+        RBFShapeFunctionsUtility::CalculateShapeFunctions(r_cloud_of_nodes_coordinates, r_slave_coordinates, r_n_container);
+
+        // Create MPCs
+        Matrix shape_matrix(1, r_n_container.size());
+        noalias(row(shape_matrix, 0)) = r_n_container; // Shape functions matrix
+        const Vector constant_vector = ZeroVector(r_n_container.size());
+        IndexType it = i.fetch_add(1); // Atomically increment the counter and get the previous value
+        concurrent_constraints.enqueue(r_clone_constraint.Create(prev_num_mpcs + it + 1, r_cloud_of_dofs, r_slave_dof, shape_matrix, constant_vector));
+    });
+
+    // Create a temporary container to store the master-slave constraints
+    ConstraintContainerType temp_constraints;
+    ModelPart::MasterSlaveConstraintType::Pointer p_constraint;
+
+    // Dequeue the constraints from the concurrent queue and add them to the temporary container
+    while (concurrent_constraints.try_dequeue(p_constraint)) {
+        temp_constraints.push_back(p_constraint);
+    }
+
+    // Add the constraints to the rComputingModelPart in a single call
+    rComputingModelPart.AddMasterSlaveConstraints(temp_constraints.begin(), temp_constraints.end());
+
+    KRATOS_INFO("AssignMPCsToNeighboursUtility") << "Build and Assign MPCs Time: " << build_and_assign_mpcs.ElapsedSeconds() << std::endl;
+    KRATOS_CATCH("");
+}
+
+void AssignMPCsToNeighboursUtility::AssignMPCsToNodes(
+    NodesContainerType pNodes,
+    double const Radius,
+    ModelPart& rComputingModelPart,
+    const Variable<array_1d<double, 3>>& rVariable,
+    double const MinNumOfNeighNodes
+    )
```

Comment:
I have a concern here, at first I though of the user passing a variable which could be either array_1d or double.
If array_1d, then you have to create the variable components lists, etc. If double, you can directly retrieve the varible with GetDof. Now I am thinking to always create a list of variables, e.g, if ["VELOCITY","PRESSURE], then create a list:
["DISPLACEMENT_X","DISPLACEMENT_Y","PRESSURE"] for 2D.
Or:
["DISPLACEMENT_X","DISPLACEMENT_Y","DISPLACEMENT_Z","PRESSURE"] for 3D.
We can discuss this in person, since it is taking a lot of time.